### PR TITLE
 SwitchCondition: Coerce left and right sides of the comparison to string

### LINKED ISSUE #1975: [Adaptive] [DCR] Fix SwitchCondition step so case values do not have to be enclosed in ''
C# implementation of SwitchCondition step requires the value property on each case to be wrapped in ''. Most users are not going to remember to do this and we do not have any validations to suggest that they need to do this and the condition just fails silently. 

Could we update the value to always be a string and automatically wrap with '' so expression evaluation goes through correctly? 

```jsonc
{
            "$type": "Microsoft.SwitchCondition",
            "condition": "dialog.listType",
            "cases": [
                {
                    "value": "todo", // today we expect value to be explicitly wrapped like 'todo'
                    "steps": [
                    ...
                    ]
                 }
}
```

### PR DESCRIPTION
**Pull request by carlosscastro**
Fixes #1975

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Steps/SwitchCondition.cs [hunk 1]
```diff
@@ -74,30 +78,45 @@ protected override async Task<DialogTurnResult> OnRunCommandAsync(DialogContext
...
+
+                    if (conditionError != null)
                     {
-                        // Do what? 
+                        throw new Exception($"Expression evaluation resulted in an error. Expression: {caseExpression.ToString()}. Error: {error}");
                     }
-                    else if (((bool)value) == true)
+
+                    // Compare both expression results. The current switch case triggers if the comparison is true.
+                    if (conditionEvaluation.ToString().Equals(value.ToString(), StringComparison.OrdinalIgnoreCase))
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Steps/SwitchCondition.cs]: > if (conditionEvaluation.ToString().Equals(value.ToString(), StringComparison.OrdinalIgnoreCase)) [](start = 19, length = 97)

this doesn't seem right.  Why not Expression.EqualsExpression(condition, engine.Parse(c.Value)) ? ?

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Steps/SwitchCondition.cs [hunk 2]
```diff
@@ -74,30 +78,45 @@ protected override async Task<DialogTurnResult> OnRunCommandAsync(DialogContext
...
+                    conditionExpression = engine.Parse(this.Condition);
                     if (this.caseExpressions == null)
                     {
                         this.caseExpressions = new Dictionary<string, Expression>();
                         foreach (var c in this.Cases)
                         {
-                            var caseCondition = Expression.EqualsExpression(condition, engine.Parse(c.Value));
-                            // map of expression to steps
+                            // Values for cases are always coerced to string
+                            var caseCondition = Expression.ConstantExpression(c.Value);
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Steps/SwitchCondition.cs]: >var caseCondition = Expression.ConstantExpression(c.Value); [](start = 28, length = 59)

this doesn't seem right, why not this?

    var caseCondition = Expression.EqualsExpression(condition, Expression.ConstantExpression(c.Value));

???

**Comment by carlosscastro** [comment]
Let me frame the issue better:

The old implementation does what you propose:

`var caseCondition = Expression.EqualsExpression(condition, engine.Parse(c.Value));`

The problem is that it treats c.Value as an expression, so 2 things happen:

1. If the switch is a string switch, you have to double quote it to tell the expression engine it is a string literal, i.e. `"case": { "value": "'frank'"}` (note double quotes around frank). Otherwise if you don't double quote it will take frank as an expression. 
2. If we change engine.Parse(c.value) to Expression.ConstantExpression(c.value) and the switch is a numeric or bool switch, this doesn't work at all. Condition is an expression for example `{ "condition": "user.age"}` and then my cases are things like `"case": { "value": "2"}`. In this case, all expressions will be false because constantExpression will consider the numbers as strings,but the condition will consider them numbers, and 2 doesn't equal '2'.

So Chris, Dong and I discussed this and there are essentially 2 ways to tackle this: make the case value an expression (more powerful, more complicated, lots of corner cases) or do what programming languages do and decide that the cases are 'compile time constants' (less powerful, more obvious for users, simple). The implication of the latter is that we coerce both sides of the equality to strings and then we compare the strings, supporting number, string and bool without double quotes required either side.

Hope this helps understand the scenario!

**Comment by tomlm** [comment]
I was thinking that you inspect the string like this
```
// if it doesn't start quote, AND it is not float AND it is not bool, then it needs to be single quoted as a string constant
if (!value.StartsWith('\'') && !Float.TryParse(value, out f) && !bool.TryParse(value, out b))
    // quote it
    value = $"'{value}'";

var valueExpression = Expression.ConstantExpression(value);
```



---
In reply to: [505951291](https://github.com/microsoft/botbuilder-dotnet/pull/2126#issuecomment-505951291) [](ancestors = 505951291)

**Comment by carlosscastro** [comment]
But doesn't that feel hackier than just letting our expression engine evaluate both sides and then checking string equality? Both are hacky in a way though so I agree that neither is great. But I don't see the upside on adding string and type manipulation. Is there any advantage I'm missing over our proposal?

**Pull request closed without merging**